block/gossip.go

```diff
@@ -4,34 +4,31 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/tendermint/tendermint/libs/pubsub"
+
 	"github.com/dymensionxyz/dymint/p2p"
 	"github.com/dymensionxyz/dymint/types"
-	"github.com/tendermint/tendermint/libs/pubsub"
 )
 
 // onNewGossipedBlock will take a block and apply it
 func (m *Manager) onNewGossipedBlock(event pubsub.Message) {
 	eventData, _ := event.Data().(p2p.GossipedBlock)
 	block := eventData.Block
 	commit := eventData.Commit
-	m.retrieverMu.Lock() // needed to protect blockCache access
-	_, found := m.blockCache[block.Header.Height]
-	// It is not strictly necessary to return early, for correctness, but doing so helps us avoid mutex pressure and unnecessary repeated attempts to apply cached blocks
-	if found {
-		m.retrieverMu.Unlock()
+	height := block.Header.Height
+
+	if m.blockCache.HasBlockInCache(height) {
 		return
 	}
 
-	m.logger.Debug("Received new block via gossip.", "block height", block.Header.Height, "store height", m.State.Height(), "n cachedBlocks", len(m.blockCache))
+	types.LastReceivedP2PHeightGauge.Set(float64(height))
+
+	m.logger.Debug("Received new block via gossip.", "block height", height, "store height", m.State.Height(), "n cachedBlocks", m.blockCache.Size())
 
 	nextHeight := m.State.NextHeight()
-	if block.Header.Height >= nextHeight {
-		m.blockCache[block.Header.Height] = CachedBlock{
-			Block:  &block,
-			Commit: &commit,
-		}
+	if height >= nextHeight {
+		m.blockCache.AddBlockToCache(height, &block, &commit)
```

Comment:
seems to me that blocks can go to the cache and never get deleted if nextHeight increases in another thread

Reply:
removed block cache mutex